packages/form-js-viewer/src/render/components/form-fields/Number.js

```diff
@@ -55,94 +54,120 @@ export function Numberfield(props) {
 
-  // checks whether the value currently in the form data is practically different from the one in the input field cache
-  // this allows us to guarantee the field always displays valid form data, but without auto-simplifying values like 1.000 to 1
-  const cacheValueMatchesState = useMemo(() => Numberfield.config.sanitizeValue({ value, formField: field }) === Numberfield.config.sanitizeValue({ value: stringValueCache, formField: field }), [ stringValueCache, value, field ]);
+  const sanitize = useCallback((value) => Numberfield.config.sanitizeValue({ value, formField: field }), [ field ]);
 
-  const displayValue = useMemo(() => {
+  const [ debouncedOnChange, flushOnChange ] = useFlushDebounce(props.onChange);
 
-    if (value === 'NaN') return 'NaN';
-    if (stringValueCache === '-') return '-';
-    return cacheValueMatchesState ? stringValueCache : ((value || value === 0) ? Big(value).toFixed() : '');
+  const previousCachedValue = usePrevious(value);
 
-  }, [ stringValueCache, value, cacheValueMatchesState ]);
-
-  const arrowIncrementValue = useMemo(() => {
-
-    if (incrementValue) return Big(incrementValue);
-    if (decimalDigits) return Big(`1e-${decimalDigits}`);
-    return Big('1');
+  useEffect(() => {
+    if (previousCachedValue !== cachedValue) {
+      debouncedOnChange({ field, value: cachedValue });
+    }
+  }, [ debouncedOnChange, cachedValue, field, previousCachedValue ]);
 
-  }, [ decimalDigits, incrementValue ]);
+  const onInputBlur = () => {
+    flushOnChange && flushOnChange();
+    onBlur && onBlur();
+  };
 
+  const onInputFocus = () => {
+    onFocus && onFocus();
+  };
 
+  // all value changes must go through this function
   const setValue = useCallback((stringValue) => {
 
     if (isNullEquivalentValue(stringValue)) {
-      setStringValueCache('');
-      onChange({ field, value: null });
+      setDisplayValue('');
+      setCachedValue(null);
       return;
     }
 
-    // treat commas as dots
+    // converts automatically for countries where the comma is used as a decimal separator
     stringValue = stringValue.replaceAll(',', '.');
 
     if (stringValue === '-') {
-      setStringValueCache('-');
+      setDisplayValue('-');
       return;
     }
 
+    // provides feedback for invalid numbers entered via pasting as opposed to just ignoring the paste
     if (isNaN(Number(stringValue))) {
-      setStringValueCache('NaN');
-      onChange({ field, value: 'NaN' });
+      setDisplayValue('NaN');
+      setCachedValue(null);
       return;
     }
 
-    setStringValueCache(stringValue);
-    onChange({ field, value: serializeToString ? stringValue : Number(stringValue) });
+    setDisplayValue(stringValue);
+    setCachedValue(sanitize(stringValue));
+
+  }, [ sanitize ]);
+
+  // when external changes occur independently of the input, we update the display and cache values of the component
+
+  const previousValue = usePrevious(value);
+
+  useEffect(() => {
```

Comment:
Same as above. Don't know if you need this wrapped in `useEffect`

Reply:
Yeah probably not, but then again it does encapsulate the behavior in a nice block. I mean either way works for me, I've tested without the effects and it's fine as well. This is more of a style thing, since executing directly in the render is basically an effect that runs every time. In terms of efficiency I'd say no effect is better, since we're just doing a single check as opposed to the multi checks involved in effect eval. But that's not something we should factor in.

Reply:
I'll follow the react guidelines on this one, and we'll go without the effects. Not particularly opinionated on this.